### PR TITLE
Relative Path Support for the Git Fetcher

### LINKED ISSUE #1237: Allow multiple profiles in a git repository
At the moment, InSpec's dependency management allows only one profile per git repository. This allows tags to be used to version the profile in the repo.

This feature request wants the ability to store multiple profiles in a single repository, under a sub-directory or at root.

Reasons that come to my mind:
- per repo pricing
- grouping of rarely updated profiles

---

Berkshelf example:

``` ruby
cookbook "mykb", git: "https://github.com/example/super_cookbooks.git", rel: "base/linux"
```

### LINKED ISSUE #1974: Add sub-directory property to Git fetcher
### Description

In my experience, various users have issues with large numbers of closely related repositories being combined into a reference platform.

See https://github.com/chris-rock/acme-inspec-profile/blob/master/inspec.yml#L10-L21 for a small example.

It would be nice to be able to specify named subdirectories for those projects that wish to use a single repository for InSpec content.

Example:

```yaml
depends:
  - name: my_inspec
    url: https://github.com/me/my_inspec_bundle
    directory: special_sauce
```

This would point `my_inspec` at the `special_sauce` subdirectory of `my_inspec_bundle` and improve the user experience.

### LINKED ISSUE #2167: Support profiles not at root level of a git repository
Please add the ability to specify a relative path when using the `git` fetcher such that InSpec can support using a profile in a repo where the repo is not at the root directory, such as a Chef cookbook used by Test Kitchen, or a repository that contains multiple profiles.

(The above is an updated description of this feature request.  Below is the original description.)

----

This was discussed briefly in https://github.com/chef/kitchen-inspec/issues/66 and determined to be a new issue.

When including a profile from an `inspec.yml` file, InSpec expects the included profile to live in top-level directories in the referenced git repository/tarball/path.

When writing a profile to be used with TestKitchen, specifically with kitchen-inspec, the profile must live in a subdirectory called `test/integration/$suite/inspec`.

Because of this discrepancy, it's impossible to write a profile that can be run with TestKitchen and also exported for use by other profiles.

### PR DESCRIPTION
**Pull request by clintoncwolfe**
## Description

This is an adoption of #3522. Main changes in the adoption were to convert the functional tests to use a fixture-based approach (instead of generating the git repos and profiles each time) and expansion of the tests.

~~WORK IN PROGRESS: there is a bug in interactions with the cache which I am currently looking at.~~ 

OF NOTE: This change requires testing against at least one local git repo. Here, this was handled by creating local git repos, then renaming the .git directory to git-fixture, creating a setup task to make and teardown a symlink, and finally I deleted all the sample hooks. One could imagine other ways of doing that.

NOTE ON PR SIZE: The vast majority of this is just test fixture code.  Main areas to focus on for human eyes are lib/inspec/fetcher/git.rb and test/functional/git_fetcher_test.rb .

Original PR below:
---

This adds support for specifying a path to a profile in a Git repo.

For example, you could have a Git repo containing multiple profiles and you want to specify a single one like below:

```yaml
depends:
  - name: profile-1
    git: https://github.com/myorg/all-my-profiles.git
    relative_path: profiles/my-specific-profile-1
  - name: profile-2
    git: https://github.com/myorg/all-my-profiles.git
    relative_path: profiles/my-specific-profile-2
```

This closes #2167 
This closes #1974 
This closes #1237

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New content (non-breaking change)
- [ ] Breaking change (a content change which would break existing functionality or processes)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.

### CODE REVIEW

File: lib/fetchers/git.rb [hunk 1]
```diff
@@ -71,7 +71,14 @@ def fetch(destination_path)
             Inspec::Log.debug("Checkout of #{resolved_ref} successful. " \
                               "Moving #{@relative_path} to #{destination_path}")
             unless File.exist?("#{working_dir}/#{@relative_path}")
-              raise ArgumentError.new("Cannot find relative path '#{@relative_path}' within profile in git repo specified by '#{@remote_url}'")
+              # Cleanup the destination path - otherwise we'll have an empty dir
+              # in the cache, which is enough to confuse the cache reader
+              # This is acourtesy, assuming we're writing to the cache; if we're
```
Comment (zenspider) [inline lib/fetchers/git.rb]: `s/acourtesy/a courtesy/`

File: lib/fetchers/git.rb [hunk 2]
```diff
@@ -71,7 +71,14 @@ def fetch(destination_path)
...
+              # vendoring to somthing more complex, don't bother.
```
Comment (zenspider) [inline lib/fetchers/git.rb]: `s/somthing/something/`

File: test/functional/git_fetcher_test.rb [hunk 3]
```diff
@@ -114,6 +114,19 @@ def assert_relative_fetch_works(profile_name, expected_profiles, expected_contro
     end
   end
 
-  # describe "running a profile with a relative path dependency that does not exist" do
-
+  describe "running a profile with a relative path dependency that does not exist" do
+    it "should fail gracefully" do
+      run_result = run_inspec_process("exec #{git_profiles}/relative-nonesuch")
+      assert_exit_code(1, run_result) # General user error
```
Comment (zenspider) [inline test/functional/git_fetcher_test.rb]: Please no more testing exit code before output. It is the least amount of usable information when things go wrong.

File: lib/fetchers/git.rb [hunk 4]
```diff
@@ -59,33 +58,30 @@ def expand_local_path(url_or_file_path)
...
-            target_profile = File.join(tmpdir, @relative_path)
-
-            FileUtils.cp_r(target_profile, dir)
+            FileUtils.cp_r("#{working_dir}/#{@relative_path}", destination_path)
           else
             Inspec::Log.debug("Checkout of #{resolved_ref} successful. " \
-                              "Moving checkout to #{dir}")
-            FileUtils.cp_r(tmpdir + "/.", dir)
+                              "Moving checkout to #{destination_path}")
+            FileUtils.cp_r(working_dir + "/.", destination_path)
```
Comment (zenspider) [inline lib/fetchers/git.rb:75]: Just curious, but why is "/." being tacked on? Can `working_dir` be... a symlink? Wouldn't that be resolved?

File: lib/fetchers/git.rb [hunk 5]
```diff
@@ -71,7 +71,14 @@ def fetch(destination_path)
...
             unless File.exist?("#{working_dir}/#{@relative_path}")
-              raise ArgumentError.new("Cannot find relative path '#{@relative_path}' within profile in git repo specified by '#{@remote_url}'")
+              # Cleanup the destination path - otherwise we'll have an empty dir
+              # in the cache, which is enough to confuse the cache reader
+              # This is acourtesy, assuming we're writing to the cache; if we're
+              # vendoring to somthing more complex, don't bother.
+              FileUtils.rmdir(destination_path) if Dir.empty?(destination_path)
+
+              raise ArgumentError.new("Cannot find relative path '#{@relative_path}' " \
+                                      "within profile in git repo specified by '#{@remote_url}'")
```
Comment (zenspider) [inline lib/fetchers/git.rb]:
```ruby
raise ArgumentError, "msg"
```

File: test/functional/git_fetcher_test.rb [hunk 6]
```diff
@@ -32,7 +32,24 @@
...
+    # Should know about the top-level profile and the child profile
+    assert_equal expected_profiles, (run_result.payload.json["profiles"].map { |p| p["name"] })
+
+    controls = run_result.payload.json["profiles"].map { |p| p["controls"] }.flatten.map { |c| c["id"] }.uniq
+    # Should have controls from the top-level and included child profile
+    expected_controls.each { |control| assert_includes controls, control }
+
+    # should not have controls from the profile defined at the top of the repo of the child profile
+    refute_includes controls, "red-dye"
+  end
```
Comment (zenspider) [inline test/functional/git_fetcher_test.rb:52]: nice

**Review by zenspider — CHANGES_REQUESTED** [review]
Minor requests, mostly typos.

File: lib/fetchers/git.rb (hunk 4 above)
Comment (clintoncwolfe) [reply lib/fetchers/git.rb:75]: I could not really say; that was in existing code, and my only changes were to clarify the variable names. I would note that `working_dir + '/', destination_path` carries a different meaning, and would copy working_dir itself, not the things in it which is what is intended. Anyway, it does continue to work.

working_dir won't be a symlink; it's the result of a `Dir.mktmpdir` block.

**Review by zenspider — APPROVED** [review]

**Review by miah — APPROVED** [review]

**Pull request merged**